core/scale/scale_encoder_stream.hpp

```diff
@@ -0,0 +1,227 @@
+     * @tparam It iterator type
+     * @param begin first iterator
+     * @param end last iterator
+     * @return reference to stream
+     */
+
+    /// Encoders
+    template<class It>
+    ScaleEncoderStream &put(It begin, It end) {
+      stream_.insert(stream_.end(), begin, end);
+      return *this;
+    }
+    /**
+     * @brief scale-encodes pair of values
+     * @tparam F first value type
+     * @tparam S second value type
+     * @param p pair of values to encode
+     * @return reference to stream
+     */
+    template <class F, class S>
+    ScaleEncoderStream &operator<<(const std::pair<F, S> &p) {
+      *this << p.first << p.second;
+      return *this;
+    }
+    /**
+     * @brief scale-encodes variant value
+     * @tparam T type list
+     * @param v value to encode
+     * @return reference to stream
+     */
+    template <class... T>
+    ScaleEncoderStream &operator<<(const boost::variant<T...> v) {
+      kagome::scale::variant::encodeVariant(v, *this);
+      return *this;
+    }
+    /**
+     * @brief scale-encodes collection of same time items
+     * @tparam T type of item
+     * @param c collection to encode
+     * @return reference to stream
+     */
+    template <class T>
+    ScaleEncoderStream &operator<<(const std::vector<T> &c) {
+      *this << BigInteger(c.size());
+      for (auto &&item : c) {
+        *this << item;
+      }
+      return *this;
+    }
+    /**
+     * @brief scale-encodes optional value
+     * @tparam T value type
+     * @param v value to encode
+     * @return reference to stream
+     */
+    template <class T>
+    ScaleEncoderStream &operator<<(const std::optional<T> &v) {
+      if (!v.has_value()) {
+        return putByte(0u);
+      }
+      return putByte(1u) << *v;
+    }
+    /**
+     * @brief scale-encodes std::array as sequence of bytes, not collection
+     * @tparam T array item type
+     * @tparam size array size
+     * @param v value to encode
+     * @return reference to stream
+     */
+    template <class T, size_t size>
+    ScaleEncoderStream &operator<<(const std::array<T, size> &v) {
+      put(v.begin(), v.end());
+      return *this;
+    }
+    /**
+     * @brief scale-encodes common::Blob as sequence of bytes, not collection
+     * @tparam size size of blob
+     * @param v blob to encode
+     * @return reference to stream
+     */
+    template <size_t size>
+    ScaleEncoderStream &operator<<(const common::Blob<size> & v) {
+      put(v.begin(), v.end());
+      return *this;
+    }
+    /**
+     * @brief scale-encodes common::Buffer as collection of bytes
+     * @param v buffer to encode
+     * @return reference to stream
+     */
+    ScaleEncoderStream &operator<<(const Buffer &v) {
+      return *this << v.toVector();
+    }
+
+    ScaleEncoderStream &operator<<(std::string_view sv) {
+      *this << BigInteger(sv.size());
+      return put(sv.begin(), sv.end());
```

Comment:
I think you can implement one templated method for Iterable (T has size(), begin(), end() and has `value_type`). something like this:
```C++
template <typename T>
ScaleEncoderStream& operator << (T&& t){
  *this << BigInteger(t.size());
  for (auto && c: t){
    *this << c;
  }
  return *this;
}
```

then add `using value_type = uint8_t` to Buffer and voila, it will work for all iterables, including all vectors, string_view, string and others.

Reply:
Implemented protected method encodeCollection(size_t, It&&begin, It&&end) and refactored those methods. By the way, std::array, common::Blob shouldn't be encoded as collections, only as sequence of bytes.